### ISSUE TITLE
How to access data submitted via “multipart/form-data” from the server side?

### ISSUE THREAD

**Issue by ChenyaZhang** [expressjs/node-multiparty#145]
In the front-end, I try to upload photos (Objective c)

```
- (void) uploadPhotoData {
    NSLog(@"111111");
    // NSURLSession Submit/Resubmit data
    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    // NSURL
    // NSURL *url = [NSURL URLWithString:[NSString stringWithFormat: @"http://localhost:8080/api/users/competitorAnalysisPhotoData/%@", self.userName]];
    NSURL *url = [NSURL URLWithString:@"http://localhost:8080/upload"];
    NSMutableURLRequest * urlRequest = [NSMutableURLRequest requestWithURL:url];
    // "Put" method
    // Request setup
    // [urlRequest setHTTPMethod:@"PUT"];
    [urlRequest setHTTPMethod:@"POST"];
    [urlRequest addValue:@"image/png" forHTTPHeaderField:@"Content-Type"];
    [urlRequest setCachePolicy:NSURLRequestReloadIgnoringLocalCacheData];
    [urlRequest setHTTPShouldHandleCookies:NO];
    [urlRequest setTimeoutInterval:30];
    // Set boundary and content type (boundary is a random String)
    NSString *boundaryConstant = [NSString stringWithFormat:@"---------------------------14737809831464368775746641449"];
    NSString* fileParamConstant = @"photo";
    NSString *contentType = [NSString stringWithFormat:@"multipart/form-data; boundary=%@", boundaryConstant];
    [urlRequest setValue:contentType forHTTPHeaderField: @"Content-Type"];
    // Image data array
    NSMutableArray *allImgData = [[NSMutableArray alloc] init];
    for (UIButton *photoButton in self.photoButtons) {
        UIImage *buttonImage = [photoButton imageForState:UIControlStateNormal];
        NSData *imgData = UIImagePNGRepresentation(buttonImage);
        // add to array
        [allImgData addObject:imgData];
        NSLog(@"22222");
    }
    // Append image data to body data
    NSMutableData *bodyData = [NSMutableData data];
    int i = 1;
    NSLog(@"allImgData.count: %lu", (unsigned long)allImgData.count);
    for (NSData *imgData in allImgData) {
        [bodyData appendData:[[NSString stringWithFormat:@"--%@\r\n", boundaryConstant] dataUsingEncoding:NSUTF8StringEncoding]];
        [bodyData appendData:[[NSString stringWithFormat:@"Content-Disposition: form-data; name=\"image\"; filename=\"%@%d.png\"\r\n", fileParamConstant, i] dataUsingEncoding:NSUTF8StringEncoding]];
        [bodyData appendData:[@"Content-Type: image/png\r\n\r\n" dataUsingEncoding:NSUTF8StringEncoding]];
        [bodyData appendData:imgData];
        [bodyData appendData:[[NSString stringWithFormat:@"\r\n"] dataUsingEncoding:NSUTF8StringEncoding]];
        i++;
        NSLog(@"!!!!!!!!!Upload 1 photo.");
    }
    // Append one last boundary
    [bodyData appendData:[[NSString stringWithFormat:@"--%@--\r\n", boundaryConstant] dataUsingEncoding:NSUTF8StringEncoding]];
    // setting the body of the post to the reqeust
    [urlRequest setHTTPBody:bodyData];
    // set the content-length
    NSString *postLength = [NSString stringWithFormat:@"%lu", (unsigned long)[bodyData length]];
    [urlRequest setValue:postLength forHTTPHeaderField:@"Content-Length"];
    // Start session task
    NSURLSessionUploadTask* task = [session uploadTaskWithRequest:urlRequest fromData:bodyData completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        NSLog(@"%@", response);
        if (error == nil && [(NSHTTPURLResponse*)response statusCode] < 300) {
            if (error == nil) {
                NSLog(@"Successfully submitted photo data. No error.");
            } else {
                NSLog(@"Photo data Error: %@", [error localizedDescription]);
                [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                    self.submitImageButton.image = [UIImage imageNamed:@"WrongUser"];
                }];
            }
        }
    }];
    [task resume];
}
```

In the back-end, I try to get all image data and store them in database (Node.js, Multiparty for http, later Mongodb, Mongoose for database):
```js
var multiparty = require('multiparty');
var http = require('http');
var util = require('util');

http.createServer(function(req, res) {

    if (req.url === '/upload' && req.method === 'POST') {
        // parse a file upload
        var count = 0;
        var form = new multiparty.Form();

        console.dir(req.body);

        // Errors may be emitted 
        // Note that if you are listening to 'part' events, the same error may be 
        // emitted from the `form` and the `part`. 
        form.on('error', function(err) {
            console.log('Error parsing form: ' + err.stack);
        });

        form.on('file', function(err, files) {
            console.log('files: %@', files);
            count++;
        });

        // Close emitted after form parsed 
        form.on('close', function() {
            console.log('Upload completed!');
            // res.setHeader('text/plain');
            // res.end('Received ' + count + ' files');
            console.log('Received ' + count + ' files');
        });

        form.parse(req, function(err, fields, files) {
            res.writeHead(200, {'content-type': 'text/plain'});
            res.write('received upload:\n\n');
            res.end(util.inspect({fields: fields, files: files}));
            // console.log('fields: %@', fields);
            // console.log('files: %@', files);
        });
    }

}).listen(8080);
```

Problem

I think I have already passed the image data to the server side, however I can only see a "path" of the image but no where to find its data:
```
Melindas-MacBook-Pro:test Chenya$ node test1.js
files: %@ { fieldName: 'image',
    originalFilename: 'photo1.png',
    path: '/var/folders/x4/01z7j18d7vz8rxrgcxt2w9dm0000gn/T/OMTuSjvM_Oj9OeUr44z0Mzk4.png',
    headers:
    { 'content-disposition': 'form-data; name="image"; filename="photo1.png"',
        'content-type': 'image/png' },
    size: 152501 }
files: %@ { fieldName: 'image',
    originalFilename: 'photo2.png',
    path: '/var/folders/x4/01z7j18d7vz8rxrgcxt2w9dm0000gn/T/Ks2y-Rsqjy_tRn_b8Psp5M88.png',
    headers:
    { 'content-disposition': 'form-data; name="image"; filename="photo2.png"',
        'content-type': 'image/png' },
    size: 152501 }
files: %@ { fieldName: 'image',
    originalFilename: 'photo3.png',
    path: '/var/folders/x4/01z7j18d7vz8rxrgcxt2w9dm0000gn/T/EoU5KUbZw9P65_UijfqPCs2_.png',
    headers:
    { 'content-disposition': 'form-data; name="image"; filename="photo3.png"',
        'content-type': 'image/png' },
    size: 157050 }
files: %@ { fieldName: 'image',
    originalFilename: 'photo4.png',
    path: '/var/folders/x4/01z7j18d7vz8rxrgcxt2w9dm0000gn/T/UjYeyX9f3m8kW6QArRQuk2-_.png',
    headers:
    { 'content-disposition': 'form-data; name="image"; filename="photo4.png"',
        'content-type': 'image/png' },
    size: 152501 }
Upload completed!
    Received 4 files
```

I try to go through the "multiparty" API (https://www.npmjs.com/package/multiparty) several times but still not sure how to locate the real data.

My use case: 1) Upload photos by user 2) Go back again to the upload page to see photos uploaded previuosly by the user 3) Download all images uploaded from server by admin

**Comment by dougwilson** [comment]
Hi @ChenyaZhang, since you are using the API to buffer the files to disk, the contents are located on your hard disk, in the file paths specified. If you want the contents, you have to either read them back out off the disk using Node.js `fs` APIs or use the event-based `part` API of this module.

**Comment by ChenyaZhang** [comment]
Thanks Doug! I tried to use the "part" API but all the debug msgs are not printed. It seems like the method is not even called? I put it in the place of "file" API as you can see from above and delete the "file" one as well. Still no reaction?

`// Parts are emitted when parsing the form 
    form.on('part', function(part) {
      // You *must* act on the part by reading it 
      // NOTE: if you want to ignore it, just call "part.resume()" 

      console.log('This is not called!');
     
      if (!part.filename) {
        // filename is not defined when this is a field and not a file 
        console.log('got field named ' + part.name);
        // ignore field's content 
        part.resume();
      }
     
      if (part.filename) {
        // filename is defined when this is a file 
        count++;
        console.log('got file named ' + part.name);
        // ignore file's content here 
        part.resume();
      }
     
      part.on('error', function(err) {
        console.log('error here!');
      });
    });`

**Comment by dougwilson** [comment]
Hi @ChenyaZhang, the `part` event is not called if you are passing a callback to `form.parse`, according to the documentation.

**Comment by ChenyaZhang** [comment]
Your comment is super helpful. Now the "part" works. 

According to the Doc: "'part' (part)
Emitted when a part is encountered in the request. part is a ReadableStream. It also has the following properties:

headers - the headers for this part. For example, you may be interested in content-type.
name - the field name for this part
filename - only if the part is an incoming file
byteOffset - the byte offset of this part in the request body
byteCount - assuming that this is the last part in the request, this is the size of this part in bytes. You could use this, for example, to set the Content-Length header if uploading to S3. If the part had a Content-Length header then that value is used here instead." 

I'm still confused about how I can use this 'part' API to access and store the image data. I tried to use byteOffset, but it did nothing. Then, where is my image data? (Thanks!!! I've been stuck in this for days...)

**Comment by ChenyaZhang** [comment]
It will be really great if there are some working examples to show how to use 'part' as you have said "you have to either read them back out off the disk using Node.js fs APIs **or use the event-based part API of this module**." I find the information from the doc on the use of 'part' is a little limited? I'm new to js so this can be hard to follow.

**Comment by dougwilson** [comment]
The `part` object is a ReadableStream, which is a core knowledge piece before starting to write in Node.js, so we really don't want to explain that... There are examples in https://github.com/expressjs/node-multiparty/tree/master/examples and you're welcome to make PRs to help improve the documentation or examples!
